### ISSUE TITLE
Excel 95 Cyrillic encoding problem

### ISSUE THREAD

**Issue by boltunoreh** [PHPOffice/PhpSpreadsheet#1018]
This is:

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
Excel 95 files with cyrillic text are processed fine

### What is the current behavior?
Cyrillic text can't be decoded well because if excel file doesn't contain codepage record, CP1252 charset is used by default and it can't be changed explicitly

### What are the steps to reproduce?
Step 1: find an excel 95 file with cyrillic text

Step 2:
```php
<?php

require __DIR__ . '/vendor/autoload.php';

$inputFileName = './excel95.xls';

$reader = new \PhpOffice\PhpSpreadsheet\Reader\Xls();
$spreadsheet = $reader->load($inputFileName);
$worksheet = $spreadsheet->getActiveSheet();

foreach ($worksheet->getRowIterator() as $row) {
    $cellIterator = $row->getCellIterator();
    $cellIterator->setIterateOnlyExistingCells(FALSE);
    foreach ($cellIterator as $cell) {
        echo $cell->getValue() . PHP_EOL;
    }
}
```

Step 3: you can see that output contains bad symbols instead of valid cyrillic:
```
Êîñòþì ôëèñîâûé
```

Using iconv() or mb_convert_encoding() do not help while changing default codepage to CP1251 does. But there is no way to set codepage.

### Which versions of PhpSpreadsheet and PHP are affected?
PhpSpreadsheet v. 1.7.0
PHP 5.6

**Comment by DmitrySkibitsky** [comment]
The problem is relevant at the moment.

**Comment by boltunoreh** [comment]
> The problem is relevant at the moment.

I've made a pull request to fix the issue #1484, but as far as I've never made pull requests to any 3rd party libs before, I'm not sure if it is correct or not. Also unfortunately I don't have enough time to add unit tests for this feature now, so I don't know is it acceptable or not.

Anyway you can use my [fork](https://github.com/boltunoreh/PhpSpreadsheet) until it is merged

**Comment by DmitrySkibitsky** [comment]
> > The problem is relevant at the moment.
> 
> I've made a pull request to fix the issue #1484, but as far as I've never made pull requests to any 3rd party libs before, I'm not sure if it is correct or not. Also unfortunately I don't have enough time to add unit tests for this feature now, so I don't know is it acceptable or not.
> 
> Anyway you can use my [fork](https://github.com/boltunoreh/PhpSpreadsheet) until it is merged

Thanks you.

**Comment by DmitrySkibitsky** [comment]
> > The problem is relevant at the moment.
> 
> I've made a pull request to fix the issue #1484, but as far as I've never made pull requests to any 3rd party libs before, I'm not sure if it is correct or not. Also unfortunately I don't have enough time to add unit tests for this feature now, so I don't know is it acceptable or not.
> 
> Anyway you can use my [fork](https://github.com/boltunoreh/PhpSpreadsheet) until it is merged

С вами можно как-то связаться? В телеграме, например?

**Comment by DmitrySkibitsky** [comment]
> > The problem is relevant at the moment.
> 
> I've made a pull request to fix the issue #1484, but as far as I've never made pull requests to any 3rd party libs before, I'm not sure if it is correct or not. Also unfortunately I don't have enough time to add unit tests for this feature now, so I don't know is it acceptable or not.
> 
> Anyway you can use my [fork](https://github.com/boltunoreh/PhpSpreadsheet) until it is merged

```php
clock($item['g']);
clock(mb_detect_encoding($item['g']));
//здесь ничего не изменилось и всё те же краказябры
$item['g'] = mb_convert_encoding($item['g'], 'utf-8', mb_detect_encoding($item['g'])); 
$item['f'] = mb_convert_encoding($item['f'], 'utf-8', mb_detect_encoding($item['f']));
```
![Аннотация 2020-05-22 140952](https://user-images.githubusercontent.com/20288556/82662167-0d170180-9c36-11ea-9237-f21711dc84fc.png)

У меня получается так, что оно то выставило CP1252 но после получания данных там всё равно кряказябры которые не получается в нормальные символы конвертировать. 

Этим методом я получаю данные с файла. 
```php
public static function toArray($filepath)
    {
        $spreadsheet = IOFactory::load($filepath);
        $worksheet = $spreadsheet->getActiveSheet();

        $rows = [];

        foreach ($worksheet->getRowIterator() as $row) {
            $cell_iterator = $row->getCellIterator();
            $cell_iterator->setIterateOnlyExistingCells(false);
            $cells = [];

            foreach ($cell_iterator as $key => $cell) {
                $cells[strtolower($cell->getColumn())] = $cell->getValue();
            }

            $rows[] = $cells;
        }

        return $rows;
    }

```

**Comment by DmitrySkibitsky** [comment]
> > The problem is relevant at the moment.
> 
> I've made a pull request to fix the issue #1484, but as far as I've never made pull requests to any 3rd party libs before, I'm not sure if it is correct or not. Also unfortunately I don't have enough time to add unit tests for this feature now, so I don't know is it acceptable or not.
> 
> Anyway you can use my [fork](https://github.com/boltunoreh/PhpSpreadsheet) until it is merged

А это файл с которым у меня проблемы https://dropmefiles.com/Mvhah

**Comment by DmitrySkibitsky** [comment]
> > The problem is relevant at the moment.
> 
> I've made a pull request to fix the issue #1484, but as far as I've never made pull requests to any 3rd party libs before, I'm not sure if it is correct or not. Also unfortunately I don't have enough time to add unit tests for this feature now, so I don't know is it acceptable or not.
> 
> Anyway you can use my [fork](https://github.com/boltunoreh/PhpSpreadsheet) until it is merged

Я вот так изменил код и начало работать.

![Аннотация 2020-05-22 143704](https://user-images.githubusercontent.com/20288556/82664183-d80cae00-9c39-11ea-95c2-e795cb3b7107.png)


![Аннотация 2020-05-22 143704ffff](https://user-images.githubusercontent.com/20288556/82664160-cb885580-9c39-11ea-84e6-58e8da9c590a.png)

**Comment by Torokh** [comment]
> > > Проблема актуальна на данный момент.
> > 
> > 
> > Я сделал запрос на перенос, чтобы исправить проблему [№ 1484](https://github.com/PHPOffice/PhpSpreadsheet/pull/1484) , но, поскольку я никогда раньше не делал запросы на перенос к каким-либо сторонним библиотекам, я не уверен, правильно это или нет. Также, к сожалению, у меня сейчас недостаточно времени для добавления модульных тестов для этой функции, поэтому я не знаю, приемлемо это или нет.
> > В любом случае вы можете использовать мою [вилку,](https://github.com/boltunoreh/PhpSpreadsheet) пока она не будет объединена
> 
> Я вот так изменил код и начало работать.
> 
> ![Аннотация 2020-05-22 143704](https://user-images.githubusercontent.com/20288556/82664183-d80cae00-9c39-11ea-95c2-e795cb3b7107.png)
> 
> ![Аннотация 2020-05-22 143704ffff](https://user-images.githubusercontent.com/20288556/82664160-cb885580-9c39-11ea-84e6-58e8da9c590a.png)

А в каком файле Вы это сделали ?

**Comment by boltunoreh** [comment]
> > > > Проблема актуальна на данный момент.
> > > 
> > > 
> > > Я сделал запрос на перенос, чтобы исправить проблему [№ 1484](https://github.com/PHPOffice/PhpSpreadsheet/pull/1484) , но, поскольку я никогда раньше не делал запросы на перенос к каким-либо сторонним библиотекам, я не уверен, правильно это или нет. Также, к сожалению, у меня сейчас недостаточно времени для добавления модульных тестов для этой функции, поэтому я не знаю, приемлемо это или нет.
> > > В любом случае вы можете использовать мою [вилку,](https://github.com/boltunoreh/PhpSpreadsheet) пока она не будет объединена
> > 
> > 
> > Я вот так изменил код и начало работать.
> > ![Аннотация 2020-05-22 143704](https://user-images.githubusercontent.com/20288556/82664183-d80cae00-9c39-11ea-95c2-e795cb3b7107.png)
> > ![Аннотация 2020-05-22 143704ffff](https://user-images.githubusercontent.com/20288556/82664160-cb885580-9c39-11ea-84e6-58e8da9c590a.png)
> 
> А в каком файле Вы это сделали ?

my PR was merged, so there is a setter method now, you can use it in your code explicitly:
`$reader->setCodepage('CP1251')`

**Comment by Torokh** [comment]
Спасибо!

С уважением,
Киселёв Станислав
тел: +7 (916) 767-0014 ( WhatsApp )
тел: +7 (925) 001-6712


<http://import.torokh.ru>Восстановление данных Apple Xsan рейдов
<http://xsan.torokh.ru>
Программирование на Delphi, Objective-C ( iPhone/iPad/MacOS X )
<http://www.torokh.ru>



вт, 13 окт. 2020 г. в 13:48, Pavel Alazankin <notifications@github.com>:

**Comment by Torokh** [comment]
Так же и осталось (


```
$reader      = \PhpOffice\PhpSpreadsheet\IOFactory::createReader("Xls");
$reader->setCodepage('CP1251');
$spreadsheet = $reader->load("3-ndfl.xls");
$locale = 'ru_ru';
$validLocale = \PhpOffice\PhpSpreadsheet\Settings::setLocale($locale);
$sheets      = $spreadsheet->getSheetCount();

$inn        = $_REQUEST['inn'];

for ($i = 0; $i < strlen($inn); $i++) {
    $spreadsheet->getActiveSheet()->setCellValueExplicit($innCell[$i], $inn[$i], \PhpOffice\PhpSpreadsheet\Cell\DataType::TYPE_STRING);
}

$writer = new \PhpOffice\PhpSpreadsheet\Writer\Xlsx($spreadsheet);
$writer->setPreCalculateFormulas(false);
$writer->save("example1.xls");
```

**Comment by oleibman** [comment]
Fixed by PR #1484 in June 2020.
